File: src/main/java/com/expense/tracker/domain/Transaction.java
package com.expense.tracker.domain;

import com.expense.tracker.adapter.DecimalUtil;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDateTime;

import static lombok.AccessLevel.NONE;
import static lombok.AccessLevel.PACKAGE;
import static lombok.AccessLevel.PRIVATE;

/**
 * @author <a href="kuldeepyadav7291@gmail.com">Kuldeep</a>
 */
@NoArgsConstructor(access = PRIVATE)
@Getter
@EqualsAndHashCode(of = "id", callSuper = false)
@Entity
@Table(name = "transaction")
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter
    private Integer id;
    @Enumerated
    @Getter(PACKAGE)
    private TxnType type;
    @Column(nullable = false)
    private String source;
    @Column(nullable = false)
    @Getter(NONE)
    private int amountInCents;
    @Column(name = "added_on", nullable = false)
    private LocalDateTime addedOn;

    private Transaction(String source, TxnType type, int amount, LocalDateTime addedOn) {
        this.source = source;
        this.amountInCents = amount;
        this.addedOn = addedOn;
        this.type = type;
    }

    public static Transaction of(String source, TxnType type, int amount, LocalDateTime addedOn) {
        if(amount <= 0) throw new IllegalArgumentException("Income cannot be 0 or negative");
        return new Transaction(source, type, amount, addedOn);
    }

    boolean isFromIncome() {
        return type == TxnType.INCOME;
    }

    boolean isFromSaving() {
        return type == TxnType.SAVING;
    }

    public int amount() {
        return amountInCents;
    }

    public BigDecimal amountInDecimals() {
        return DecimalUtil.centsToDecimal(amountInCents);
    }
}

File: src/main/java/com/expense/tracker/domain/Transaction.kt
package com.expense.tracker.domain

import com.expense.tracker.adapter.DecimalUtil
import java.math.BigDecimal
import java.time.LocalDateTime
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.Table

@Entity
@Table(name = "transaction")
class Transaction (
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Int? = null,
        @Column(nullable = false) val source: String,
        @Enumerated val type: TxnType,
        @Column(nullable = false) val amountInCents: Int,
        @Column(name = "added_on", nullable = false) val addedOn: LocalDateTime
) {

    fun setId(id: Int) {
        this.id = id
    }

    val isFromIncome: Boolean
        get() = type == TxnType.INCOME
    val isFromSaving: Boolean
        get() = type == TxnType.SAVING

    fun amount(): Int {
        return amountInCents
    }

    fun amountInDecimals(): BigDecimal {
        return DecimalUtil.centsToDecimal(amountInCents)
    }

    companion object {
        fun of(source: String, type: TxnType, amount: Int, addedOn: LocalDateTime): Transaction {
            require(amount > 0) { "Income cannot be 0 or negative" }
            return Transaction(null, source, type, amount, addedOn)
        }
    }
}